docs/proposals/107-spark-client/README.md

```diff
@@ -0,0 +1,1023 @@
+    executor_gpu={"name": "nvidia.com/gpu", "quantity": 2},
+
+    # Node selector and tolerations
+    node_selector={"node-type": "compute"},
+    tolerations=[{
+        "key": "spark",
+        "operator": "Equal",
+        "value": "true",
+        "effect": "NoSchedule"
+    }],
+)
+```
+
+### Backend Configurations
+
+#### OperatorBackendConfig
+
+```python
+from kubeflow.spark.backends import OperatorBackendConfig
+
+config = OperatorBackendConfig(
+    namespace="default",                     # K8s namespace
+    context=None,                            # kubeconfig context
+    service_account="spark-operator-spark",  # Service account
+    image_pull_policy="IfNotPresent",        # Image pull policy
+    default_spark_image="gcr.io/spark-operator/spark-py",
+    enable_monitoring=True,                  # Prometheus monitoring
+    enable_ui=True,                          # Spark UI
+    timeout=60,                              # API timeout (seconds)
+)
+```
+
+#### GatewayBackendConfig
+
+```python
+from kubeflow.spark.backends import GatewayBackendConfig
+from kubeflow.spark.config import AuthMethod
+
+config = GatewayBackendConfig(
+    gateway_url="http://gateway:8080",
+    user="myuser",
+    password="mypassword",
+    auth_method=AuthMethod.BASIC,
+    timeout=30,
+    verify_ssl=True,
+)
+```
+
+### Core Components
+
+#### 1. SparkBackend (Abstract Base)
+
+Defines the interface all backends must implement:
+
+```python
+class SparkBackend(abc.ABC):
+    @abc.abstractmethod
+    def submit_application(...) -> SparkApplicationResponse
+
+    @abc.abstractmethod
+    def get_status(submission_id: str) -> ApplicationStatus
+
+    @abc.abstractmethod
+    def wait_for_completion(...) -> ApplicationStatus
+
+    @abc.abstractmethod
+    def get_logs(...) -> Iterator[str]
+
+    @abc.abstractmethod
+    def list_applications(...) -> List[ApplicationStatus]
+
+    @abc.abstractmethod
+    def delete_application(submission_id: str) -> Dict
+```
+
+#### 2. OperatorBackend (Kubernetes CRD)
+
+Implements the backend using Spark Operator CRDs:
+
+- Creates `SparkApplication` CRDs with full configuration
+- Manages lifecycle through Kubernetes API
+- Supports all Spark Operator features (dynamic allocation, monitoring, GPU)
+- Provides log streaming from driver and executor pods
+- Enables Spark UI access through services/ingress
+
+#### 3. GatewayBackend (REST API)
+
+Implements the backend for managed Spark gateways:
+
+- Submits applications via REST endpoints
+- Compatible with Apache Livy and similar gateways
+- Supports basic auth, token auth, and custom authentication
+- Provides session-based and batch job execution
+
+
+#### 3. ConnectBackend (Spark Connect / Interactive Sessions)
+
+Implements session-oriented backend using Spark Connect protocol (gRPC):
+
+**Key Features**:
+- Remote connectivity to existing Spark clusters via Spark Connect (gRPC)
+- Session-based interactive workloads (exploratory analysis, notebooks)
```

Comment:
We also need to figure out whether this capability might be useful for Notebooks use-cases.
I know that @lresende and @fresende did some work to integrate Spark Operator with Jupyter Enterprise gateway: https://www.kubeflow.org/docs/components/spark-operator/user-guide/notebooks-spark-operator/

cc @ederign @andyatmiami @akshaychitneni @shravan-achar